From c1918fcb1e6a7772f3e8c23657c1fb521dce5805 Mon Sep 17 00:00:00 2001
From: Thomas Huth <thuth@redhat.com>
Date: Wed, 1 Oct 2025 09:23:06 +0200
Subject: [PATCH] Revert "hw/s390x/s390-virtio-ccw: Remove the deprecated 4.1
 machine type"

JIRA: https://issues.redhat.com/browse/RHEL-105900
Upstream status: downstream-only

This reverts commit 6ad3a47f15624176434490e895ac244375322607.

We still need the code for the old machine types since the -rhel7.6.0
machine type is based on these old upstream machine types.

Signed-off-by: Thomas Huth <thuth@redhat.com>
---
 hw/s390x/s390-virtio-ccw.c  | 15 +++++++++++++++
 target/s390x/gen-features.c |  4 ++++
 2 files changed, 19 insertions(+)

diff --git a/hw/s390x/s390-virtio-ccw.c b/hw/s390x/s390-virtio-ccw.c
index d5d32f6a75..0ff4114d2a 100644
--- a/hw/s390x/s390-virtio-ccw.c
+++ b/hw/s390x/s390-virtio-ccw.c
@@ -1167,6 +1167,21 @@ static void ccw_machine_4_2_class_options(MachineClass *mc)
     compat_props_add(mc->compat_props, hw_compat_4_2, hw_compat_4_2_len);
 }
 DEFINE_CCW_MACHINE(4, 2);
+
+static void ccw_machine_4_1_instance_options(MachineState *machine)
+{
+    static const S390FeatInit qemu_cpu_feat = { S390_FEAT_LIST_QEMU_V4_1 };
+    ccw_machine_4_2_instance_options(machine);
+    s390_set_qemu_cpu_model(0x2964, 13, 2, qemu_cpu_feat);
+}
+
+static void ccw_machine_4_1_class_options(MachineClass *mc)
+{
+    ccw_machine_4_2_class_options(mc);
+    compat_props_add(mc->compat_props, hw_compat_4_1, hw_compat_4_1_len);
+}
+DEFINE_CCW_MACHINE(4, 1);
+
 #endif /* disabled for RHEL */
 
 static void ccw_rhel_machine_9_6_0_instance_options(MachineState *machine)
diff --git a/target/s390x/gen-features.c b/target/s390x/gen-features.c
index 8218e6470e..a814ece82f 100644
--- a/target/s390x/gen-features.c
+++ b/target/s390x/gen-features.c
@@ -884,6 +884,9 @@ static uint16_t qemu_MIN[] = {
      */
     S390_FEAT_FLOATING_POINT_EXT,
     S390_FEAT_ZPCI,
+};
+
+static uint16_t qemu_V4_1[] = {
     S390_FEAT_STFLE_53,
     S390_FEAT_VECTOR,
 };
@@ -1046,6 +1049,7 @@ static FeatGroupDefSpec FeatGroupDef[] = {
  *******************************/
 static FeatGroupDefSpec QemuFeatDef[] = {
     QEMU_FEAT_INITIALIZER(MIN),
+    QEMU_FEAT_INITIALIZER(V4_1),
     QEMU_FEAT_INITIALIZER(V6_0),
     QEMU_FEAT_INITIALIZER(V6_2),
     QEMU_FEAT_INITIALIZER(V7_0),
